### DISCUSSION TITLE
Progress bar when uploading multiple files?

### DISCUSSION THREAD

**Discussion by ternaus** [fastapi/fastapi#6782] [Questions] [answered]
Following the documentation for https://fastapi.tiangolo.com/tutorial/request-files/#multiple-file-uploads to upload multiple files.

Everything works.

The question is - how can I add a progress bar to the website to show the progress of the file upload?

**Comment by Kludex** [comment, marked as answer]
I think it will be better if you could upload file by file and give the progress responsibility to the frontend. But if that's not something you want to do (or can't), I've done a small snippet on a way you can accomplish that.

The database below is illustrative, but you probably want an in-memory database (e.g. redis).

```python
from asyncio import sleep
from typing import List

from fastapi import FastAPI, File, UploadFile
from fastapi.responses import HTMLResponse
from pydantic import BaseModel

app = FastAPI()

class UploadFileProgress(BaseModel):
    total: int
    uploaded: int

database = {"upload_tasks": {}}


@app.post("/uploadfiles/")
async def create_upload_files(files: List[UploadFile] = File(...)):
    task_id = len(database["upload_tasks"])
    database["upload_tasks"] = {task_id: {"total": len(files), "uploaded": 0}}
    for i, file in enumerate(files, start=1):
        await sleep(5) # upload files
        database["upload_tasks"][task_id]["uploaded"] = i


@app.get("/progress/{task_id}/", response_model=UploadFileProgress)
async def get_upload_progress(task_id: int):
    return database["upload_tasks"][task_id]
```

**Comment by ternaus** [comment]
Thank you. Uploading file by file and dealing with the progress bar using Javascript did not cross my mind. This works.

**Comment by tiangolo** [comment]
Thanks for the help here @Kludex ! 👏  🙇 

Thanks for reporting back and closing the issue @ternaus  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.

**Comment by ScrimForever** [comment]
Thinking some option to use tqdm to show progress of upload files, but all things that i tried fail. If someone can help me or think toghether

**Reply by ScrimForever** [reply]
```
async def enviar_arquivo(file: UploadFile):
    total_size = file.size
    chunk_size = 1024
    chunks = max(1, int(math.ceil(file_size / chunk_size)))
    progress = tqdm(desc=f"Uploading", total=total_size, unit="B", unit_scale=True)
    async with aiofiles.open(file.filename, 'wb') as f:
        for _ in range(chunks):
            chunk = await f.read(chunk_size)
            progress.update(len(chunk))
            yield chunk
```

This an example that is not work
